### ISSUE TITLE
Support K8 resource YAML config as input for DSL components

### ISSUE THREAD

**Issue by swiftdiaries** [kubeflow/pipelines#756]
Part of issues related to having better support for k8s resources in Pipelines.

It'd better if we can re-use existing implementations for creating individual components.
For example, TFJob and most CRDs in Kubeflow already have a ksonnet prototype defined. If we could use these as input to create components, it'd be really cool.
Another use case is when you want to have features from Argo that are not yet implemented in Pipelines DSL. Like the docker-in-docker workflows for Seldon: [YAML](https://github.com/kubeflow/example-seldon/blob/master/workflows/serving-tf-mnist-workflow.yaml)

Related issue: #677 
Related discussion: https://kubeflow.slack.com/archives/CE10KS9M4/p1548777905172300

**Comment by Ark-kun** [comment]
CRs are somewhat limited as components.
Components should have inputs and outputs while CRs are pretty limited here.
The way both Pipelines (see kubeflow TFJob launcher) and Argo currently handle CRs is by using a controlling container.
We might probably provide an easier way to submit and wait for CRs in future, but the problems with inputs and outputs remains.

**Comment by Ark-kun** [comment]
Does this do what you want to accomplish?

```
import kfp.components as comp

create_seldon_deployment_op = comp.load_component_from_text(
'''
name: Create Seldon deployment
implementation:
  container:
    image: lachlanevenson/k8s-kubectl
    command: [sh, -c, 'echo "$0" | kubectl apply -f -']
    args:
    - |
      apiVersion: "machinelearning.seldon.io/v1alpha1"
      kind: "SeldonDeployment"
      ...
''')

def pipeline():
    create_seldon_deployment_task = create_seldon_deployment_op()
```

**Comment by vicaire** [comment]
@swiftdiaries, would something like this work: 

1) At the low level, we provide a generic container that takes a CRD spec as a parameter and runs it to completion. 

2) The DSL can implement some special, easier to use abstraction on top of that on a case by case basis, for the CRDs that are used the most. 

3) For debugging, the UI should be able to view/follow the tree of K8 resources (container labels indicated the Argo workflow that created them, custom resource labels indicate which containers created them)

**Comment by vicaire** [comment]
Note that ksonnet is being archived (https://ksonnet.io/) so we probably won't spend time integrating the DSL with ksonnet.

**Comment by elikatsis** [comment]
@swiftdiaries, Hello!

It's been some time, so let's see what we've got and what more do we need to close this issue.

Regarding the volume usage you mention in [this comment](https://github.com/kubeflow/pipelines/issues/756#issuecomment-460164174):
* You've implemented the solution you proposed in #783
* Also, there are `pvolumes` semantics:
  * `ContainerOp` constructor argument introduced by #926
  * `add_pvolumes()` method introduced by #1353

As far as the K8s resources are concerned, which is the main reason of this issue, I think it is covered by `ResourceOp`s (introduced by #926). These implement Argo's resource template, for K8s resource manipulation and there's logging by Argo.

Finally, #879, which introduced sidecars, enabled the use of full K8s specifications through its client for containers.

So, do these cover this issue? Is there something I'm missing?

**Comment by swiftdiaries** [comment]
@elikatsis I think this can be closed :) 
Thanks for the ping !
